src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/ReadAndUpdateOperation.java

```diff
@@ -26,47 +28,94 @@ public record ReadAndUpdateOperation(
     boolean returnDocumentInResponse,
     boolean returnUpdatedDocument,
     boolean upsert,
-    Shredder shredder)
+    Shredder shredder,
+    int updateLimit)
     implements ModifyOperation {
 
   @Override
   public Uni<Supplier<CommandResult>> execute(QueryExecutor queryExecutor) {
-    Uni<ReadOperation.FindResponse> docsToUpate = readOperation().getDocuments(queryExecutor, null);
+    final AtomicBoolean moreDataFlag = new AtomicBoolean(false);
+    final Multi<ReadOperation.FindResponse> findResponses =
+        Multi.createBy()
+            .repeating()
+            .uni(
+                () -> new AtomicReference<String>(null),
+                stateRef -> {
+                  Uni<ReadOperation.FindResponse> docsToDelete =
+                      readOperation().getDocuments(queryExecutor, stateRef.get());
+                  return docsToDelete
+                      .onItem()
+                      .invoke(findResponse -> stateRef.set(findResponse.pagingState()));
+                })
+            .whilst(findResponse -> findResponse.pagingState() != null);
+    final AtomicInteger matchedCount = new AtomicInteger(0);
+    final AtomicInteger modifiedCount = new AtomicInteger(0);
+
     final Uni<List<UpdatedDocument>> updatedDocuments =
-        docsToUpate
+        findResponses
             .onItem()
             .transformToMulti(
                 findResponse -> {
-                  if (findResponse.docs().isEmpty()) {
-                    if (upsert) {
-                      return Multi.createFrom().item(readOperation().getEmptyDocuments());
+                  final List<ReadDocument> docs = findResponse.docs();
+                  if (upsert() && docs.size() == 0 && matchedCount.get() == 0) {
+                    return Multi.createFrom().item(readOperation().getEmptyDocuments());
```

Comment:
It sounds like this would be resulting in document upsert? But mention of "empty document" would suggest something else -- or is that just a document with `_id` field and nothing else?

Reply:
The method just creates an empty document, if _id filter is part of filter it adds it too.

Reply:
Will rename to newDocument()